lib/phoenix/presence.ex

```diff
@@ -421,4 +430,64 @@ defmodule Phoenix.Presence do
       end)
     end)
   end
+
+  @doc """
+  Keep a presence state in sync.
+    * `presence_list` - A list of presences that was obtained by calling `Presence.list/1`
+    * `diff` - The `payload` from a `presence_diff` broadcast.
+  ## Examples
+      iex> state = %{
+          u1: %{metas: [%{id: 1, phx_ref: "1"}]},
+          u2: %{metas: [%{id: 2, phx_ref: "2"}]},
+          u3: %{metas: [%{id: 3, phx_ref: "3"}]}
+        }
+      iex> diff = %{
+        joins: %{u1: %{metas: [%{id: 1, phx_ref: "1.2"}]}},
+        leaves: %{u2: %{metas: [%{id: 2, phx_ref: "2"}]}}
+      }
+      iex> Phoenix.Presence.sync_diff(state, diff)
+      %{
+        u1: %{metas: [%{id: 1, phx_ref: "1"}, %{id: 1, phx_ref: "1.2"}]},
+        u3: %{metas: [%{id: 3, phx_ref: "3"}]}
+      }
+  """
+
+  def sync_diff(state, %{joins: joins, leaves: leaves}) do
+    Enum.reduce(leaves, state, fn leave, new_state ->
+      delete_meta_or_presence(new_state, leave)
+    end)
+    |> Map.merge(joins, fn _k, state_presence, joined_presence ->
+      %{metas: add_or_replace_metas(state_presence.metas, joined_presence.metas)}
+    end)
+  end
+
+  defp delete_meta_or_presence(state, {key, %{metas: left_metas}}) do
+    state_metas = Map.get(state, key, %{}) |> Map.get(:metas, [])
+    remaining_metas = state_metas -- left_metas
+
+    case remaining_metas do
+      [] -> Map.delete(state, key)
+      _ -> Map.put(state, key, %{metas: remaining_metas})
+    end
+  end
+
+  defp add_or_replace_metas(state_metas, joined_metas) do
+    Enum.filter(state_metas, fn state_meta ->
+      keep_existing_meta?(state_meta, joined_metas)
```

Comment:
@josevalim you mentioned _"discard joins of entries that we already know"_

This happens inside of a `Map.merge/3`'s 3rd argument function. So it might feel like it is not very performant but this code only runs when there is a conflict, and the lists of metas should only be as big as how many devices a single user signed in from.

Before discarding anything, we need to be sure that it is the exact same entry that we already know.

It felt easier to discard our existing entries that match the incoming entry`phx_ref` at the same time as discarding entries that might not be identical but need to be replaced because they match the `phx_ref_prev` and then `++ joined_metas`.